src/Microsoft.AspNetCore.Razor.Evolution/Legacy/ParserVisitor.cs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.Evolution.Legacy
+{
+    internal abstract class ParserVisitor
+    {
+        public virtual void VisitBlock(Block block)
+        {
+            VisitStartBlock(block);
+            
+            for (var i = 0; i < block.Children.Count; i++)
+            {
+                block.Children[i].Accept(this);
+            }
+
+            VisitEndBlock(block);
+        }
+
+        public virtual void VisitStartBlock(Block block)
```

Comment:
Seems wasteful/overly complex to have all these start/end mechanics given how we're using them (to create a container object of a specific IR Node type).

Why not have each of the call points be responsible for diving their children? Aka:
```C#
Builder.Push(new CustomIRNodeType() { ... })
visitor.VisitBlock(block); // Dive children
Builder.Pop();
```

Could also take a block creation approach where the call to accept is expected to output some container type that's used as a parent to dive the children.

Lots of choices available. Just really don't like start/end 😄